utils/src/priority_set.rs

```diff
@@ -0,0 +1,152 @@
+use std::cmp::Ordering;
+use std::collections::{BTreeSet, HashMap, HashSet};
+use std::hash::Hash;
+
+#[derive(Eq, PartialEq, Clone)]
+struct Entry<I: Ord + Hash + Clone, V: Ord + Clone> {
+    item: I,
+    value: V,
+}
+
+impl<I: Ord + Hash + Clone, V: Ord + Clone> Ord for Entry<I, V> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.value.cmp(&other.value) {
+            Ordering::Equal => self.item.cmp(&other.item),
+            other => other,
+        }
+    }
+}
+
+impl<I: Ord + Hash + Clone, V: Ord + Clone> PartialOrd for Entry<I, V> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A set that offers fast, priority-ordered iteration over
+/// its elements.
+pub struct PrioritySet<I: Ord + Hash + Clone, V: Ord + Clone> {
```

Comment:
The alternative approach to this (to reduce the copying on `put`) would be to use an indexed heap.